### PR TITLE
unr-636: Make generated schema IDs stable between runs of the schema generator

### PR DESCRIPTION
**Pull request by chrisfields0**
#### Description
code changes relating to unr-636
- implemented a feature to keep generated schema persistent and not destructive in between runs.
- implemented ability to load classes into memory as long as they've already had schema generated for the class atleast once.
- deletes garbage/unused data from SchemaDatabase if a class no longer exists within the project
#### Tests
Tested on ShooterGame:
1. Load Highrise
2. click Generate Schema 
3. click Generate Snapshot
4. Load ShooterEntry
5. click Generate Schema
6. click Generate Snapshot
7. Load Highrise again
8. Generate Snapshot
9 Launch SpatialOS and Play. 

#### Documentation
https://improbableio.atlassian.net/browse/UNR-636

#### Primary reviewers
@m-samiec @davedolben

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SchemaDatabase.cpp [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#include "Utils/SchemaDatabase.h"
+
+USchemaDatabase::USchemaDatabase() : NextAvailableComponentId(SpatialConstants::STARTING_GENERATED_COMPONENT_ID)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SchemaDatabase.cpp]: If this is all that's here, you could just default initialize the variable inline.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 2]
```diff
@@ -152,15 +154,40 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
 {
-	Worker_ComponentId ComponentId = SpatialConstants::STARTING_GENERATED_COMPONENT_ID;
-
 	// Generate the actual schema
-	for (auto& TypeInfo : TypeInfos)
+	for (const auto& TypeInfo : TypeInfos)
 	{
-		ComponentId += GenerateCompleteSchemaFromClass(CombinedSchemaPath, ComponentId, TypeInfo);
+		const UClass* const Class = Cast<UClass>(TypeInfo->Type);
+		const FSchemaData* const SchemaData = ClassPathToSchema.Find(*Class->GetPathName());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Move this declaration inside if statement condition.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 3]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+		ClassPathToSchema = SchemaDatabase->ClassPathToSchema;
+		NextAvailableComponentId = SchemaDatabase->NextAvailableComponentId;
+	}
+	else
+	{
+		// As a safety precaution, if the SchemaDatabase.uasset doesn't exist then make sure the schema generated folder is cleared as well. 
+		DeleteGeneratedSchemaFiles();
+	}
+}
+#pragma optimize("", off)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Remove #pragma

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 4]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+void PreProcessSchemaMap()
+{
+	TArray<FString> EntriesToRemove;
+	for (const auto& EntryIn : ClassPathToSchema)
+	{
+		UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Checking if following file exists: %s"), *(EntryIn.Key));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Change to verbose

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 5]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+
+		const FSoftObjectPath ItemToReference(EntryIn.Key);
+
+		// first check if the object is already loaded into memory
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
		// First check if the object is already loaded into memory.
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 6]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+	{
+		// As a safety precaution, if the SchemaDatabase.uasset doesn't exist then make sure the schema generated folder is cleared as well. 
+		DeleteGeneratedSchemaFiles();
+	}
+}
+#pragma optimize("", off)
+void PreProcessSchemaMap()
+{
+	TArray<FString> EntriesToRemove;
+	for (const auto& EntryIn : ClassPathToSchema)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: `EntryIn` should probably be called `ClassPath`

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 7]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+	{
+		UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Checking if following file exists: %s"), *(EntryIn.Key));
+
+		const FSoftObjectPath ItemToReference(EntryIn.Key);
+
+		// first check if the object is already loaded into memory
+		UObject* const ResolvedObject = ItemToReference.ResolveObject();
+		UObject* const LoadedObject   = ResolvedObject ? nullptr : ItemToReference.TryLoad();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: We only care about UClass objects here, so may as well cast immediately.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 8]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+
+		// only store classes that weren't currently loaded into memory
+		if (LoadedObject && !AdditionalSchemaGeneratedClasses.Find(Cast<UClass>(LoadedObject)))
+		{
+			// don't allow the Garbage Collector to delete these objects
+			LoadedObject->AddToRoot();
+			AdditionalSchemaGeneratedClasses.Add(Cast<UClass>(LoadedObject));
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Delete line

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 9]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+		}
+
+		// if the class isn't loaded then mark the entry for removal from the map
+		else if(!ResolvedObject && !LoadedObject)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Can just check LoadedObject here. And LoadedObject != nullptr

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 10]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+	}
+}
+#pragma optimize("", off)
+void PreProcessSchemaMap()
+{
+	TArray<FString> EntriesToRemove;
+	for (const auto& EntryIn : ClassPathToSchema)
+	{
+		UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Checking if following file exists: %s"), *(EntryIn.Key));
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Define a FString variable `ClassPath`, set to EntryIn.Key, and use it throughout.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 11]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+{
+	// gets the classes currently loaded into memory
+	SchemaGeneratedClasses = GetAllSupportedClasses();
 	SchemaGeneratedClasses.Sort();
 
 	// Generate Type Info structs for all classes
 	TArray<TSharedPtr<FUnrealType>> TypeInfos;
 
-	for (const auto& Class : SchemaGeneratedClasses)
+	for ( UClass*  Class : SchemaGeneratedClasses)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Why the change?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 12]
```diff
@@ -244,18 +272,84 @@ TArray<UClass*> GetAllSupportedClasses()
...
+
+		// first check if the object is already loaded into memory
+		UObject* const ResolvedObject = ItemToReference.ResolveObject();
+		UObject* const LoadedObject   = ResolvedObject ? nullptr : ItemToReference.TryLoad();
+
+		// only store classes that weren't currently loaded into memory
+		if (LoadedObject && !AdditionalSchemaGeneratedClasses.Find(Cast<UClass>(LoadedObject)))
+		{
+			// don't allow the Garbage Collector to delete these objects
+			LoadedObject->AddToRoot();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Worth adding a note that these objects will *never* get cleaned up, even if moving between maps.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 13]
```diff
@@ -277,21 +371,12 @@ bool SpatialGDKGenerateSchema()
...
-		PlatformFile.CreateDirectory(*SchemaOutputPath);
-	}
-
 	check(GetDefault<UGeneralProjectSettings>()->bSpatialNetworking);
-	GenerateSchemaFromClasses(TypeInfos, SchemaOutputPath);
 
-	SaveSchemaDatabase();
+	DeleteGeneratedSchemaFiles();
+
+	SaveSchemaDatabase(GenerateSchemaFromClasses(TypeInfos, SchemaOutputPath));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Break this line into two calls.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 14]
```diff
@@ -24,7 +24,9 @@
 DEFINE_LOG_CATEGORY(LogSpatialGDKSchemaGenerator);
 
 TArray<UClass*> SchemaGeneratedClasses;
+TArray<UClass*> AdditionalSchemaGeneratedClasses;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Add some comments about what these arrays are used for.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 15]
```diff
@@ -152,15 +154,40 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
 }
 }// ::
 
-void GenerateSchemaFromClasses(const TArray<TSharedPtr<FUnrealType>>& TypeInfos, const FString& CombinedSchemaPath)
+const Worker_ComponentId GenerateSchemaFromClasses(const TArray<TSharedPtr<FUnrealType>>& TypeInfos, const FString& CombinedSchemaPath)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
Worker_ComponentId GenerateSchemaFromClasses(const TArray<TSharedPtr<FUnrealType>>& TypeInfos, const FString& CombinedSchemaPath)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 16]
```diff
@@ -152,15 +154,40 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+		const UClass* const Class = Cast<UClass>(TypeInfo->Type);
+		const FSchemaData* const SchemaData = ClassPathToSchema.Find(*Class->GetPathName());
+		
+		if (SchemaData)
+		{
+			bool bFoundNonZeroId = false;
+
+			for (const auto& CurrComponentId : SchemaData->SchemaComponents)
+			{
+				if (CurrComponentId)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
				if (CurrComponentId > 0)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SchemaDatabase.cpp (hunk 1 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SchemaDatabase.cpp]: deleted the SchemaDatabase.cpp file

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 4 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Ah, I actually didn't mean to keep that log statement in there. Do you think it should be but with verbose flag?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Nah, fine to delete.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 11 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: this was accidental.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 7 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: changed.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 13 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: changed.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 10 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: changed.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp [hunk 17]
```diff
@@ -313,13 +314,15 @@ int GenerateActorSchema(int ComponentId, UClass* Class, TSharedPtr<FUnrealType>
...
 		}
 
+		const Worker_ComponentId CachedComponentId = SchemaData ? SchemaData->SchemaComponents[PropertyGroupToSchemaComponentType(Group)] : 0;
+
 		Writer.PrintNewLine();
 
 		Writer.Printf("component {0} {", *SchemaReplicatedDataName(Group, Class));
 		Writer.Indent();
-		Writer.Printf("id = {0};", IdGenerator.GetNextAvailableId());
+		Writer.Printf("id = {0};", CachedComponentId > 0 ? CachedComponentId : IdGenerator.GetNextAvailableId());
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp]: The pattern of `CachedComponentId > 0 ? CachedComponentId : IdGenerator.GetNextAvailableId()` appears about 8 times. Could this be condensed maybe? Maybe making the IdGenerator knowledgeable about cached component ids?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 18]
```diff
@@ -244,12 +247,78 @@ TArray<UClass*> GetAllSupportedClasses()
...
-bool SpatialGDKGenerateSchema()
+void DeleteGeneratedSchemaFiles()
+{
+	const FString SchemaOutputPath = GetDefault<USpatialGDKEditorSettings>()->GetGeneratedSchemaOutputFolder();
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.DirectoryExists(*SchemaOutputPath))
+	{
+		if (!PlatformFile.DeleteDirectoryRecursively(*SchemaOutputPath))
+		{
+			UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Could not clean the generated schema directory '%s'! Please make sure the directory and the files inside are writeable."), *SchemaOutputPath);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Should this be a log or maybe a popup?

Comment (Vatyx) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Eh nevermind, a log is fine.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 19]
```diff
@@ -244,12 +247,78 @@ TArray<UClass*> GetAllSupportedClasses()
...
-	SchemaGeneratedClasses = GetAllSupportedClasses();
+	if (SchemaDatabase)
+	{
+		ClassPathToSchema = SchemaDatabase->ClassPathToSchema;
+		NextAvailableComponentId = SchemaDatabase->NextAvailableComponentId;
+	}
+	else
+	{
+		NextAvailableComponentId = SpatialConstants::STARTING_GENERATED_COMPONENT_ID;
+		// As a safety precaution, if the SchemaDatabase.uasset doesn't exist then make sure the schema generated folder is cleared as well.
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Can you log here that no SchemaDatabase is was found so we will be clearing out the old folder (if it exists) and generating a new one?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/ComponentIdGenerator.h [hunk 20]
```diff
@@ -4,18 +4,19 @@
...
-	FComponentIdGenerator(int StartId) : InitialId(StartId), NumIds(0)
+	FComponentIdGenerator(int StartId) : InitialId(StartId), IdReturned(StartId), NumIds(0)
 	{
 	}
 
-	int GetNextAvailableId()
+	int GetNextAvailableId(const uint32 InComponentId)
 	{
-		return InitialId + (NumIds++);
+		IdReturned = InComponentId > 0 ? InComponentId : InitialId + (NumIds++);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/ComponentIdGenerator.h]: InComponentId != SpatialConstants::INVALID_COMPONENT_ID

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp [hunk 21]
```diff
@@ -313,11 +314,13 @@ int GenerateActorSchema(int ComponentId, UClass* Class, TSharedPtr<FUnrealType>
 			continue;
 		}
 
+		const Worker_ComponentId CachedComponentId = SchemaData ? SchemaData->SchemaComponents[PropertyGroupToSchemaComponentType(Group)] : 0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp]: Use SpatialConstants::INVALID_COMPONENT_ID

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/ComponentIdGenerator.h [hunk 22]
```diff
@@ -4,18 +4,19 @@
...
+		return IdReturned;
 	}
 
 	int GetCurrentId()
 	{
-		return InitialId + (NumIds - 1);
+		return IdReturned;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/ComponentIdGenerator.h]: Driveby: this function should be const

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 23]
```diff
@@ -24,7 +24,9 @@
 DEFINE_LOG_CATEGORY(LogSpatialGDKSchemaGenerator);
 
 TArray<UClass*> SchemaGeneratedClasses;
+TArray<UClass*> AdditionalSchemaGeneratedClasses; // used to load and keep objects into memory
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Perhaps, `Used to keep UClasses in memory whilst generating schema for them.` ?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 24]
```diff
@@ -244,12 +247,78 @@ TArray<UClass*> GetAllSupportedClasses()
...
+	for (const auto& EntryIn : ClassPathToSchema)
+	{
+		const FString ClassPath = EntryIn.Key;
+		const FSoftObjectPath ItemToReference(ClassPath);
+		
+		// First check if the object is already loaded into memory.
+		UObject* const ResolvedObject = ItemToReference.ResolveObject();
+		UClass*  const LoadedClass    = ResolvedObject ? nullptr : Cast<UClass>(ItemToReference.TryLoad());
+		
+		const bool bUniqueEntry = AdditionalSchemaGeneratedClasses.Find(LoadedClass) == INDEX_NONE;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Was this ever not true? Shouldn't ClassPathToSchema only contain unique values?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 25]
```diff
@@ -277,21 +346,22 @@ bool SpatialGDKGenerateSchema()
...
-
 	check(GetDefault<UGeneralProjectSettings>()->bSpatialNetworking);
+
+	DeleteGeneratedSchemaFiles();
+
 	GenerateSchemaFromClasses(TypeInfos, SchemaOutputPath);
 
 	SaveSchemaDatabase();
 
+	for (const auto& EntryIn : AdditionalSchemaGeneratedClasses)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Add a log about what you're doing here and why.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 24 above)
Comment (chrisfields0) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: No, it was always returning true because the TArray::Find func return -1 if it doesn't find the item. -1, when converted to a boolean, returns true.

The new statement fixes this issue and also ensures that AdditionalSchemaGeneratedClasses only contains unique values.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: My question is why would we be adding duplicate values to the array? What's feeding it should contain unique values.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 26]
```diff
@@ -277,21 +345,24 @@ bool SpatialGDKGenerateSchema()
...
-	if (PlatformFile.DirectoryExists(*SchemaOutputPath))
+	check(GetDefault<UGeneralProjectSettings>()->bSpatialNetworking);
+
+	DeleteGeneratedSchemaFiles();
+
+	GenerateSchemaFromClasses(TypeInfos, SchemaOutputPath);
+
+	SaveSchemaDatabase();
+
+	//Allow the garbage collector to clean up classes that were manually loaded and forced to keep alive for the Schema Generator process.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]:
```suggestion
	// Allow the garbage collector to clean up classes that were manually loaded and forced to keep alive for the Schema Generator process.
```
